### PR TITLE
feat: oneshot channel returns `Pkid`

### LINKED ISSUE #349: Get packet id for publish
There doesn't seem to be a way to get the packet id for a published QoS1/2 message, to track its state
Is this correct? Is there a workaround?

### PR DESCRIPTION
**Pull request by de-sh**
Fixes #349, implements #805 

## Type of change

New feature (non-breaking change which adds functionality) 

Allows user access to pkid information once it is assigned by rumqttc internals during run.

## Checklist:

- [x] Formatted with `cargo fmt`
- [x] Make an entry to `CHANGELOG.md` if it's relevant to the users of the library. If it's not relevant mention why.

### Tests performed
Ran this against provided example async code, output looks as follows
```
Event = Incoming(ConnAck(ConnAck { session_present: false, code: Success }))
Event = Outgoing(Subscribe(1))
Pkid: 1
Event = Outgoing(Publish(2))
Pkid: 2
Event = Incoming(SubAck(SubAck { pkid: 1, return_codes: [Success(AtMostOnce)] }))
Event = Outgoing(PubRel(2))
Event = Incoming(PubRec(PubRec { pkid: 2 }))
Event = Incoming(Publish(Topic = hello/world, Qos = AtMostOnce, Retain = false, Pkid = 0, Payload Size = 1))
Event = Incoming(PubComp(PubComp { pkid: 2 }))
Event = Outgoing(Publish(3))
Pkid: 3
Event = Outgoing(PubRel(3))
Event = Incoming(PubRec(PubRec { pkid: 3 }))
Event = Incoming(Publish(Topic = hello/world, Qos = AtMostOnce, Retain = false, Pkid = 0, Payload Size = 2))
Event = Incoming(PubComp(PubComp { pkid: 3 }))
Event = Outgoing(Publish(4))
Pkid: 4
Event = Outgoing(PubRel(4))
Event = Incoming(PubRec(PubRec { pkid: 4 }))
Event = Incoming(Publish(Topic = hello/world, Qos = AtMostOnce, Retain = false, Pkid = 0, Payload Size = 3))
Event = Incoming(PubComp(PubComp { pkid: 4 }))
Event = Outgoing(Publish(5))
Pkid: 5
Event = Outgoing(PubRel(5))
Event = Incoming(PubRec(PubRec { pkid: 5 }))
Event = Incoming(Publish(Topic = hello/world, Qos = AtMostOnce, Retain = false, Pkid = 0, Payload Size = 4))
Event = Incoming(PubComp(PubComp { pkid: 5 }))
Event = Outgoing(Publish(6))
Pkid: 6
Event = Outgoing(PubRel(6))
Event = Incoming(PubRec(PubRec { pkid: 6 }))
Event = Incoming(Publish(Topic = hello/world, Qos = AtMostOnce, Retain = false, Pkid = 0, Payload Size = 5))
Event = Incoming(PubComp(PubComp { pkid: 6 }))
Event = Outgoing(PingReq)
Event = Outgoing(Publish(7))
Pkid: 7
Event = Incoming(PingResp)
Event = Outgoing(PubRel(7))
Event = Incoming(PubRec(PubRec { pkid: 7 }))
Event = Incoming(Publish(Topic = hello/world, Qos = AtMostOnce, Retain = false, Pkid = 0, Payload Size = 6))
Event = Incoming(PubComp(PubComp { pkid: 7 }))
Event = Outgoing(Publish(8))
Pkid: 8
Event = Outgoing(PubRel(8))
Event = Incoming(PubRec(PubRec { pkid: 8 }))
Event = Incoming(Publish(Topic = hello/world, Qos = AtMostOnce, Retain = false, Pkid = 0, Payload Size = 7))
Event = Incoming(PubComp(PubComp { pkid: 8 }))
Event = Outgoing(Publish(9))
Pkid: 9
Event = Outgoing(PubRel(9))
Event = Incoming(PubRec(PubRec { pkid: 9 }))
Event = Incoming(Publish(Topic = hello/world, Qos = AtMostOnce, Retain = false, Pkid = 0, Payload Size = 8))
Event = Incoming(PubComp(PubComp { pkid: 9 }))
Event = Outgoing(Publish(10))
Pkid: 10
Event = Outgoing(PubRel(10))
Event = Incoming(PubRec(PubRec { pkid: 10 }))
Event = Incoming(Publish(Topic = hello/world, Qos = AtMostOnce, Retain = false, Pkid = 0, Payload Size = 9))
Event = Incoming(PubComp(PubComp { pkid: 10 }))
Event = Outgoing(Publish(11))
Pkid: 11
Event = Outgoing(PubRel(11))
Event = Incoming(PubRec(PubRec { pkid: 11 }))
Event = Incoming(Publish(Topic = hello/world, Qos = AtMostOnce, Retain = false, Pkid = 0, Payload Size = 10))
Event = Incoming(PubComp(PubComp { pkid: 11 }))
Event = Outgoing(PingReq)
Event = Incoming(PingResp)
```
Thus the implementation is complete.

### CODE REVIEW

**Review by swanandx — COMMENTED** [review]
I like the idea and also tried it out, with both sync & async, and it works nicely!

Here is my feedback:
- we should avoid placing the `pkid_tx` directly in packets, mqttbytes should strictly be related to protocol parsing. We can maybe put it in request like `Request::Publish(tx, publish)` ?
- just `Pkid` seems intuitive than `PkidPromise` 
- users don't need to know about the oneshot channel, thus instead of just type alias, we should impl / rename some methods. e.g. async client is fine, but when in sync, using `pkid.blocking_recv()` isn't. maybe we can rename that recv to something else?
- if we fail to _send_ the pkid from state, maybe we can log some debug lvl info? not sure how often we will fail though.
- examples can be simplified, we just want to show _how user can get pkid_. So we can simply `await` right after getting `Pkid`. Using joinsets / delay queues in unnecessarily adding complexity. Also, users want to associate the publish with the pkid right? awaiting em later via joinset or something will be same as checking for `Outgoing` event. 
- as you suggested in RFC, having many open oneshot channels might addup perf impact which we are unsure of as of now. Whether to feature gate this or not is your call ( as you would know better haha )

**Comment by de-sh** [comment]
> as you suggested in RFC, having many open oneshot channels might addup perf impact which we are unsure of as of now. Whether to feature gate this or not is your call ( as you would know better haha )

File: rumqttc/src/v5/state.rs [hunk 1]
```diff
@@ -183,21 +186,21 @@ impl MqttState {
     /// be put on to the network by the eventloop
     pub fn handle_outgoing_packet(&mut self, request: Request) -> Result<(), StateError> {
         match request {
-            Request::Publish(publish) => {
+            Request::Publish(tx, publish) => {
                 self.check_size(publish.size())?;
```
Comment (de-sh) [inline rumqttc/src/v5/state.rs:190]: `EventLoop` drops the `pkid_tx` here if size is not valid

**Comment by de-sh** [comment]
I have a few comments about the `PkidError` type, it needs to be more descriptive of why it failed to resolve into a valid number. This could be one of many:
1. `EventLoop` was shutdown, hence the tx was dropped.
2. `EventLoop` rejected the packet and hence did not assign it a pkid

We should probably handle this in a separate PR

**Comment by swanandx** [comment]
> it's just to give users an idea of what is possible, they can choose to not go the route we illustrate, but we can try and educate them about the possibilities.
>
my thinking was, the examples should _show_ them _how_ they can get the pkid in general. for simple examples sake, getting pkid one by one should be fine imo ( in async & in sync )

we can make separate examples tailored toward why they might want to use joinset ( or something else ). There we should also explain why it is used and when they should and shouldn't use it.

> From a user perspective, you wouldn't want them to confuse between the awaiting and the occupied state.
>
I didn't get this completely, but i would like to differ. Adding `Promise` to name is too verbose(? idk correct word). e.g. the receiver is `Receiver` not `ReceiverPromise`.

**Comment by de-sh** [comment]
> e.g. the receiver is Receiver not ReceiverPromise.

**Pull request closed without merging**